internal/service/image/image.go

```
@@ -0,0 +1,116 @@
+package image
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/distribution/reference"
+)
+
+const (
+	LatestTag = "latest"
+	MainTag   = "main"
+)
+
+var (
+	ErrEmptyImageURL       = errors.New("empty image URL")
+	ErrImageNameExtraction = errors.New("could not extract image name")
+	ErrNoTagOrDigest       = errors.New("no tag or digest found")
+	ErrMissingImageTag     = errors.New("image is missing a tag")
+	ErrDisallowedTag       = errors.New("image tag is disallowed (latest/main)")
+)
+
+type ImageInfo struct {
+	Name    string
+	Tag     string
+	Digest  string
+	FullURL string
+}
+
+func IsValidImage(value string) (bool, error) {
```

This function has a strange signature, maybe I am confused because it's not documented...
What's the point of returning `false, nil`?
I can see there's some "advanced machinery" involved in actually parsing the image and producing meaningful description of what went wrong.
But in some cases we silently skip the "what is wrong" - we're just saying "it's wrong" without providing any explanation... Personally, I would return a meaningful error  - always.
This would make the function simpler to use, you just have two cases (err == nil vs err != nil) instead of three: (false + nil, false + not-nil, true + nil)

In summary:
- If returning (false, nil) is required, document the contract for this particular outcome
- Consider returning just error. No error => valid.

---

I understand the pragmatic approach of fast, "coarse" filtering first, and more expensive, fine-grained checks later.
But you can still return an error for "fast/coarse" filtering - it's enough to return error instead of `false`
In this way:
- you have simpler signature
- you don't have to write this documentation. The code is self-documenting via errors.

About performance: Yes, this WILL BE less performant, but notice by how much:
The additional cost involved is to allocate a single error.
Is that much?
If the performance of creating a single error value is a performance concern, then probably the entire docker parsing function would be considered a no-go here and we should switch to a really low-level code that returns integers to indicate error conditions.
I don't think we're facing that much of a perormance issue here.